src/transformStateWithClones.js

```diff
@@ -1,13 +1,33 @@
-'use strict';
-
-/**
- * @param {Object} state
- * @param {Object[]} actions
- *
- * @return {Object[]}
- */
+"use strict";
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultArray = [{ ...state }];
+
+  for (let i = 0; i < actions.length; i++) {
+    const valueBeforePassing =
+      i === 0 ? { ...resultArray[i] } : { ...resultArray[i - 1] };
+
+    switch (actions[i].type) {
+      case "addProperties":
+        const addProperties = actions[i].extraData;
+
+        resultArray[i] = Object.assign(valueBeforePassing, addProperties);
+        break;
+      case "removeProperties":
+        const arrayElementToDeleted = actions[i].keysToRemove;
+
+        arrayElementToDeleted.forEach((removeElement) => {
+          delete valueBeforePassing[removeElement];
+        });
+        resultArray[i] = valueBeforePassing;
```

Comment:
Instead of deleting key-value pairs one by one in a loop, you can use reduce function which is more efficient and cleaner.